### ISSUE TITLE
Remote authentication using environmental variables does not work

### ISSUE THREAD

**Issue by seppestas** [conan-io/conan#10280]
Authenticating using the [CONAN_LOGIN_USERNAME, CONAN_LOGIN_USERNAME_{REMOTE_NAME}](https://docs.conan.io/en/latest/reference/env_vars.html#conan-login-username-conan-login-username-remote-name) and [CONAN_PASSWORD, CONAN_PASSWORD_{REMOTE_NAME}](https://docs.conan.io/en/latest/reference/env_vars.html#conan-password-conan-password-remote-name) environmental variables does not work. At least searching for packages and installing them does not work with this authentication method in Windows cmd and Powershell and it appears to not work either on Linux with Bash (Docker container based on Ubuntu 18.04).

### Environment Details (include every applicable attribute)
  * Operating System+version: Windows 10 Build 19042.1348
  * Compiler+version: N/A
  * Conan version: 1.44.0
  * Python version: 3.8.6

### Steps to reproduce (Include if Applicable)

In CMD:

```cmd
conan user --clean
set CONAN_LOGIN_USERNAME=<user>
set CONAN_PASSWORD=<password>
conan search -r <remote> <package>
```

Powershell:

```psh
conan user --clean
$env:CONAN_LOGIN_USERNAME="<user>"
$env:CONAN_PASSWORD="<password>"
conan search -r <remote> <package>
```

No packages are found. Error message: "There are no packages matching the '<package>' pattern". Setting a breakpoint at [_get_env_password](https://github.com/conan-io/conan/blob/f230bf104842c160be6ea377497151fe1ea64c38/conans/client/userio.py#L103) it appears the environmental variables are never even loaded.

The user command can be used successfully with the same login and password:

CMD:

``` 
conan user %CONAN_LOGIN_USERNAME% -p %CONAN_PASSWORD%
conan search -r <remote> <package>
```

PSH:

```
conan user $env:CONAN_LOGIN_USERNAME -p $env:CONAN_PASSWORD
conan search -r <remote> <package>
```

Output:

```
Existing package recipes:

<package>/5.5.3@qorvo/beta
<package>/5.5.6@qorvo/release
<package>/5.5.8@qorvo/master
<package>/5.5.9@qorvo/master
<package>/5.8.0@qorvo/release
<package>/6.0.0@qorvo/release
<package>/6.0.1@qorvo/release
...
```

Related tickets:

- #7793
- #5434

This missing feature makes installing Conan packages in a CI/CD environment more tricky than it should be. As mentioned by the Conan documentation "These variables are useful for unattended executions like CI servers or automated tasks". Also note e.g [Gitlab CI specifically documents using these environmental variables to authenticate iof the user command](https://docs.gitlab.com/ee/user/packages/conan_repository/index.html#authenticate-to-the-package-registry).

**Comment by Ext3h** [comment]
The problem with `CONAN_LOGIN_USERNAME` is the conditions necessary for it to be evaluated at all.

Let's start with the obvious first, the tests forgot to cover for anything *except* upload: https://github.com/conan-io/conan/blob/83d13540abbefd1bbcfcd8d6c03e4f6d570dbeed/conans/test/integration/remote/auth_test.py#L59

So why *does* it work for uploads? Because that API endpoint usually triggers an `AuthException`, and only the occurrence of that exception triggers the authentication using this environment variable: https://github.com/conan-io/conan/blob/173cbc6143133c83ed1ecb9b5315d3ca787a846a/conans/client/rest/auth_manager.py#L44

How to fix? Well, don't wait for  an exception to occur, but authorize *eagerly* in the presence of this variable.

For any request hitting the search endpoints, you won't get any exception. You can't, given how the conan repositories are usually structured. They do have to provide different packages for different user groups, so the search APIs *can't* tell you a restrictive "some possible result was restricted for anon". but are required to give you the *least* restrictive response which is *we couldn't give you any results" or "partial results only".

**Comment by memsharded** [comment]
Hi @Ext3h 

Thanks for your feedback.

The whole auth system is based on the efficiency of caching a JWT. It is not good to do an https basic auth call before every other API call if the CONAN_PASSWORD is defined

In any case, this is solely a responsibility of the server. If the server decides to give a valid, empty response for a repo, when credentials are not provided, instead of asking for auth, it is their choice. Client applications do not need to pre-emptively log-in or submit credentials into servers before they are requested, just in case the server will not be responding the whole information.

I think you might want to have a look to the server side configuration in order to request auth, it works also for the ``conan search``, I have just tried:

```bash
(conan36) λ conan remote add myremote https://xxxxxxxxxxxxxxxxx.jfrog.io/artifactory/api/conan/conan-local

$ conan search * -r=myremote
Please log in to "myremote" to perform this action. Execute "conan user" command.
Remote 'myremote' username:
```

**Comment by seppestas** [comment]
> The whole auth system is based on the efficiency of caching a JWT

Shouldn't it also be easy to evaluate the expiry time of a JWT? Caching is great, as long as there's a good system for cache invalidation.

In my case I was using Giltab's Conan server. Sadly (speaking from memory) it does not seem to return an expiry time in the JWTs, but it has a hard-coded, if I remember correctly rather short, expiry time for the tokens.

A good sollution could be to:
1. Check if an expiry time is set in the JWT (might already be done, I think I saw the v2 protocol support something like this)
2. Allow to set a hardcoded expiry time per repository that can be used with the iat time if no ext time is set in the JWT.

If the server decides to give a valid, empty response for a repo, when credentials are not provided, instead of asking for auth, it is their choice.

Agreed, but an understandable one. This prevents information from being leaked. An other way of thinking about this is "there are no packages in this repository for you (the unauthenticated user)" iof "you need to be authenticated to list the packages".

> In any case, this is solely a responsibility of the server.

I don't agree. The Conan CLI should err on the side of caution. If a JWT exists without expiry time, it should have a way of invalidating it.

> It is not good to do an https basic auth call before every other API call if the CONAN_PASSWORD is defined

I understand that this in not preferable if you want to interactively use Conan, but for server to server communication in CI/CD I think this is actually preferred so things like state (the cached tokens) don't complicate things.

At the very least, this behavior should be documented. Again, the Conan docs have this to say:
> You can define the username for the authentication process using environment variables. Conan will use a variable CONAN_LOGIN_USERNAME_{REMOTE_NAME}, if the variable is not declared Conan will use the variable CONAN_LOGIN_USERNAME, if the variable is not declared either, Conan will request to the user to input a username.

> These variables are useful for unattended executions like CI servers or automated tasks.

However, this is clearly not the case. A `conan remote login` command has to be used to unsure the JWT is created and cached for subsequent calls.

**Comment by memsharded** [comment]
Sorry, but I disagree too :)

It is not the responsibility of the client either to check expiration of tokens. The token can be expired by the server for many other reasons, and the token declare expire date is irrelevant from the client side, it cannot just expect that the token will be ok if the expiration date is ok.
It is the responsibility of the server to launch an auth response to the client if the token has expired, so the client auths again. In this case, the Conan client will use the CONAN_XXX env-vars to re-authenticate and get a new token. This should be already working.

**Comment by seppestas** [comment]
> It is the responsibility of the server to launch an auth response to the client if the token has expired

Good point. There is a difference between no auth and expired auth. Sounds like a bug for Gitlab's Conan server implementation.

Either way, the Conan client will not use the CONAN_XXX env vars if *no* token exists. At least not at the time I was testing it (and based on @Ext3h's feedback this seems to still be the case). I'm no longer using Conan so I can't really confirm.

Note in my original post, I explicitly call `conan user --clean`, so I expect Conan to ignore any previous tokens. I would consider it a bug if it even attempted to use old tokens after this command, because it might use an incorrect user (you could even argue Conan should check if the cached token is for the same user, see #7793).

**Comment by Ext3h** [comment]
> I think you might want to have a look to the server side configuration in order to request auth, it works also for the conan search, I have just tried:

There's a catch - this only works for the simplest of possible setups. When you choose to do the entire stack of access management on a per-repository base, and require the user to register every single repository with the client. Use in enterprise setups doesn't give you 1 or 2 repos. It gives you dozens. Per Artifactory instance (including caches for other sites).

We stumbled across this with the use of *virtual* repositories in Artifactory, which hide the complexity of access control to like 30-40 different repositories by providing the user only a single API endpoint for read access. The way the conan API is designed, this endpoint has to behave different depending on which (or if any) credentials are used. Only for commands such as `upload` - which are non-ambiguously inaccessible for anonymous users, *forcing* authentication is even possible.

> It is the responsibility of the server to launch an auth response to the client if the token has expired, so the client auths again.

Artifactory does that. But as stated, it won't do that as long as anonymous access has any sane behavior defined to it, and you are accessing as anonymous user by the default. The search APIs are just such a special case, they just return empty sets (well - 404 actually).

This was easy enough to resolve with a simple `conan user $env:CONAN_LOGIN_USERNAME -p $env:CONAN_PASSWORD` - but as it stands the documentation is straight up lying when it indicates that `CONAN_LOGIN_USERNAME` could be used as a drop-in for storing the credentials explicitly with `conan user`.

And yeah, the fact that conan will then still cache the resulting *token and the user name* and a `conan user --clean` is necessary regardless to clear the authorization from the config doesn't behave at all as you would expected from an environment variable either.

**Comment by memsharded** [comment]
Ok, I think we are not going to fully agree on this.. :)

I still think that the current approach is the best one (taking into account all trade-offs), and that it is not possible to start checking credentials, authorizing in the servers, etc, because a ``conan search`` is going to request the list of packages. At least not automatically, built-in. I think this could have quite a few rough edges and unexpected side effects, that would make worse something that has been working quite robustly for thousands of enterprises. I still appreciate that there might be some cases like yours (part of it seems actually not related to Conan, but to Gitlab server implementation), that something could be improved, or at least make the UI smoother, so this is what I would suggest:

- Improve the documentation, to make it clearer the general behavior
- To work on a explicit ``conan remote auth -r=*`` (or ``conan remote list-users --auth``, something in that idea), that would automate this task: checking if current tokens if existing are valid, if not or not existing checking the CONAN_XXX env-vars and try to auth with them, and finishing with a report. 

That would probably improve enough the UX, just one simple command, while leaving not affected other users that are fine with the current behavior, and having enough flexibility to control the repos to auth into.

**Comment by seppestas** [comment]
Maybe just remove the CONAN_LOGIN_* environment variables, since arguably they don't work as intended? Conan remote login works well enough and is only slightly less convenient than the env vars.

Next step is to get Gitlab's Conan server to return the proper error if the token expires.

**Comment by memsharded** [comment]
> Maybe just remove the CONAN_LOGIN_* environment variables, since arguably they don't work as intended? Conan remote login works well enough and is only slightly less convenient than the env vars.

I agree, but yours and mine "slightly less convenient", for others is a big deal. These env-vars were introduced because for many users, doing the explicit auth in their pipelines was too inconvenient, they have to repeat it in every pipeline (while the env-var is defined at the CI level), but the worse was the risk of leaking credentials and having them exposed in the jobs logs, because whatever logging strategy they had that was logging every command literally, including the env-vars values replaced in command line expansion 🤦‍♂️ . For these reasons I think I'd prefer not to remove them at this moment, and probably work on the new convenience "remote auth" command.

I'll roll this discussion to the team, to get extra feedback too. Thanks!

**Comment by Ext3h** [comment]
For starters, just adjusting the documentation to warn of the pitfalls should do.

Namely: Beware that these variables won't behave even remotely as expect when accessing any API which shows byzantine behavior between anonymous and fully authorized access, and link directly back to  `conan user` command as the way to go whenever possible.

Removing is too late, I fully agree on that.

**Comment by seppestas** [comment]
> These env-vars were introduced because for many users, doing the explicit auth in their pipelines was too inconvenient, they have to repeat it in every pipeline

Well, this is exactly what I was trying to do, but it doesn't work as expected, at least not when installing Conan dependencies, which is way more inconvinient IMO. I would prefer the feature working as intended, but IMO it's better not to have the feature than to have it only work partly and confuse users.

If I run a Gitlab CI job that tries to install Conan dependencies from a Gitlab Conan repo, it does not work if I set the CONAN_LOGIN_* env vars.

I get the argument that we expect the repository to return an unauthenticated error if we try to fetch packages without authenticating, but I don't think that is correct. If no token exists, Conan should, as @Ext3h stated, "authorize eagerly". We derailed a bit before on the whole JWT expiring, but this is not relevant in this case.

Even if this feature works, e.g for Conan repos that do return an unauthenticated / unauthorized error when trying to list packages without an auth token, I would argue this could cause issues if the user set in the env vars is different than what is cached in the token.
